doc/specs/stdlib_selection.md

```diff
@@ -0,0 +1,364 @@
+to find a smaller or larger rank (that will have led to partial sorting of
+`index`, thus implying some constraint on the location).
+
+#### Licensing
+
+The Fortran Standard Library is distributed under the MIT
+License. However components of the library may be based on code with
+additional licensing restrictions. In particular `select` and `arg_select`
+were derived by modifying a matlab implementation of "qselect" by Manolis
+Lourakis, https://www.mathworks.com/matlabcentral/fileexchange/68947-qselect
+Below is the licence of the matlab qselect
+
+Copyright (c) 2018, Manolis Lourakis
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+* Redistributions of source code must retain the above copyright notice, this
+  list of conditions and the following disclaimer.
+
+* Redistributions in binary form must reproduce the above copyright notice,
+  this list of conditions and the following disclaimer in the documentation
+  and/or other materials provided with the distribution
+* Neither the name of Foundation for Research and Technology - Hellas nor the
+  names of its contributors may be used to endorse or promote products
+  derived from this software without specific prior written permission.
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+### Specifications of the `stdlib_selection` procedures
+
+#### `select` - find the kth smallest value in an input array
+
+##### Status
+
+Experimental
+
+##### Description
+
+Returns the k-th smallest value of array `a(:)`, and also partially sorts `a(:)`
+such that `all(a(1:k) <= a(k))`  and `all(a(k) <= a((k+1):size(a)))`
+
+##### Syntax
+
+`call [[stdlib_selection(module):select(interface)]]( a, k, kth_smallest [, left, right ] )`
+
+##### Class
+
+Generic subroutine.
+
+##### Arguments
+
+`a` : shall be a rank one array of any of the types:
+`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`,
+`real(sp)`, `real(dp)`, `real(qp)`. It is an `intent(inout)` argument. On input it is
+the array in which we search for the kth smallest entry. On return its elements
+will be partially sorted such that:
+ `all( a(1:k-1) <= a(k) )` and `all(a(k) <= a(k+1:size(a)))`.
```

Comment:
I suggest to move the sentence `On input ... a)))`. to the next section.

Reply:
I'm not sure about this - do you mean move it to the "Notes" section?

One thing I am conscious of is that some new users will be confused by the fact that `array` is changed, and I really want to make that point "hard to miss" in the documentation.

We could potentially just say `... On outputs its elements are reordered, see Notes below`. But I'm not sure if that is better than the current approach. Thoughts?

Reply:
You mentioned it is an `intent(inout)`. So it is clear enough for me.

I think that the sentence I pointed out should be moved to the `result/output` section.

Reply:
I implemented this in some updates just now.